synapse/handlers/device.py

```diff
@@ -495,13 +495,11 @@ async def notify_device_update(
                 "Notifying about update %r/%r, ID: %r", user_id, device_id, position
             )
 
-        room_ids = await self.store.get_rooms_for_user(user_id)
-
         # specify the user ID too since the user should always get their own device list
         # updates, even if they aren't in any rooms.
-        self.notifier.on_new_event(
-            "device_list_key", position, users=[user_id], rooms=room_ids
-        )
+        users_to_notify = users_who_share_room.union(user_id)
```

Comment:
I'm a bit confused by this, it doesn't seem to just be re-using something but passing in a totally different value?

I think this is wrong though, `.union(user_id)` will cause each character to be iterated and added separately to the result.

Reply:
> I'm a bit confused by this, it doesn't seem to just be re-using something but passing in a totally different value?

We are creating a new value, but through copying rather than recalculating it.

`on_new_event` eventually needs to work out which users to send the updates to. Doing this by deriving a list of users from a list of room IDs is expensive. We already have an appropriate list of users to notify in `users_who_share_room`, so this PR proposes to just use that.

I've just realised another angle to this change. Device list updates are separate from rooms, and thus shouldn't end up passing a list of room IDs to `on_new_event`, but a list of users. For instance, if an application service explicitly put a room's ID in its room ID namespace (effectively subscribing to updates from the room), it shouldn't then receive device list updates from users in that room. It should only get them if a user in the room falls within its user namespace, as device list updates are between users.

> I think this is wrong though, `.union(user_id)` will cause each character to be iterated and added separately to the result.

Ah, that is a silly mistake. I was surprised CI didn't catch it though. I've written a test to do so in the future: https://github.com/matrix-org/synapse/pull/11909.

Reply:
Over chat @anoadragon453 explained to me that this fixes a case of local users not having their sync streams woken up when users in the same room have new devices (that was my interpretation at least).

Do we need to test for this behavior? Are there any currently open issues about this? (I thought we had one, but can't find it on looking again.)

Reply:
> this fixes a case of local users not having their sync streams woken up when users in the same room have new devices (that was my interpretation at least).

Correct interpretation! (Un)fortunately, my assertion was not correct. `/sync` *does* end up populating the `room_to_user_streams` dict of `Notifier` on each `/sync`. This hidden bit in the call stack does it:

https://github.com/matrix-org/synapse/blob/64ec45fc1b0856dc7daacca7d3ab75d50bd89f84/synapse/notifier.py#L487-L488

Having the appropriate entry in `room_to_user_streams` means that passing a room ID to `Notifier.on_new_event` will correctly translate to, and thus wake up, the associated user stream. And lo, `/sync` streams do get woken up correctly in this case.

Understanding this did take writing a unit test to really prove though - and hooray for more tests! Tests have been put up separately in https://github.com/matrix-org/synapse/pull/11972.

So what remains for this PR? It doesn't fix a bug by passing users to `on_new_event`; `room_ids` works perfectly fine. The only benefit at this point is to save recalculating the list of users from the given `room_ids` before we eventually call `notify_interested_services_ephemeral` at the end of `on_new_events`.

Now that I have a better understanding of how the notifier code works, I've realised that calling `notify_interested_services_ephemeral` in `on_new_event` is a bit of a hack. It's probably better to wrap this method up into a Notifier callback. We'd need to set up user streams on appservice registration though... a topic for another issue.

For now, a solution is to have `notify_interested_services_ephemeral` accept a collection of room IDs and use those to deduce which users to notify. It is extra effort over what this PR proposes, but I'm not convinced that this PR is the Right Way To Do Things:tm: anymore.

So! Closing this, and will reopen with a change that doesn't influence other (nuanced) components so much.

Reply:
Thanks for the explanation! It sounds like this code is confusingly entangled with other bits. 😢

Reply:
Actually actually, on further reflection, creating a user stream for every AS user would be much more overhead than simply calculating which AS's to send an event to when it happens.

So... the current code isn't really a hack. In fact, it's probably the best way to do it. But the AS-related code in `Notifier` code use a refactor to make different paths better isolated from each other.